### ISSUE TITLE
Running OZ against mainnet fork overwrites `.openzeppelin/mainnet.json`

### ISSUE THREAD

**Issue by asselstine** [OpenZeppelin/openzeppelin-sdk#1306]
I'm using ganache-cli to fork mainnet and test contract upgrades.

I wanted to keep the fork as separate as possible, so I updated the OZ networks like so:

```js
module.exports = {
  networks: {
    mainnet: {
      provider: () => new HDWalletProvider(
        process.env.HDWALLET_MNEMONIC,
        process.env.INFURA_PROVIDER_URL_MAINNET
      )
    },

    mainnet_fork: {
      provider: () => new HDWalletProvider(
        process.env.HDWALLET_MNEMONIC,
        process.env.LOCALHOST_URL // using ganache-cli --fork
      )
    }
  }
}
```

Now I can simply redirect OZ using `--network mainnet_fork`.

However, the problem is the OZ still writes out any changes to `.openzeppelin/mainnet.json`.  This is a problem!  It means I have to be super diligent about undoing any changes made to this file before committing with Git.

I've read the documentation regarding the [`<network>.json`](https://github.com/OpenZeppelin/openzeppelin-sdk/blob/00d15b997d76e6f439895afa0f584235c40c65aa/packages/docs/modules/ROOT/pages/configuration.adoc) file, so I understand the intention behind this behaviour.  However, how can we take this use case into account?  Forking mainnet is crucial to test upgrades.

Perhaps we can add a "name" attribute to the network that can override the filename?

Something like:

```js
module.exports = {
  networks: {
    mainnet: {
      provider: () => new HDWalletProvider(
        process.env.HDWALLET_MNEMONIC,
        process.env.INFURA_PROVIDER_URL_MAINNET
      )
    },

    mainnet_fork: {
      name: 'mainnet_fork',
      provider: () => new HDWalletProvider(
        process.env.HDWALLET_MNEMONIC,
        process.env.LOCALHOST_URL // using ganache-cli --fork
      )
    }
  }
}
```

This would preserve the existing behaviour, but allow people like myself to determine the filename.

Thoughts?

**Comment by nventuro** [comment]
Hello @asselstine! This is a very interesting scenario, thanks for sharing!

By looking at `ganache-cli`'s documentation, it looks like you should be able to call it with the `--networkId` option on a forked network, and have a different id that way. By doing this, The SDK will no longer think your fork and mainnet are the same network, and you'll get different entries on your project file. Would this solve your issue, or do you need the network id to be the same?

Do note that you won't have a project file for the forked network, so you'll have to copy it yourself from the mainnet entry, but that shouldn't be hard. 

We're currently discussing how to improve the user experience of testing an upgrade on a live system, so your feedback is extremely valuable to us. If you've had any other issues while performing this task, or have ideas on ways in which we could make the process easier, please share them!

**Comment by asselstine** [comment]
Interesting- I didn't see that option.  I just tried it and it does the trick!  Now I have another `dev-*.json` network config.  Perfect.  Thanks @nventuro 

Copying the network file is not a big deal, but it might be worth thinking about whether it's useful for OZ support forking at a deeper level.  Perhaps it could automate some of this work.

In my case, I've added a simple 'yarn fork' command that will spin up a ganache-cli fork.  It's custom because I need certain user accounts unlocked.

I imagine there could be a `forks.js` file that defines forking rules for each network.  OZ users could setup the unlocked accounts list.  Any network could be forked, and OZ commands could be run against that fork.

If you want to chat about it, feel free to reach out!  Fork testing is essential so first-class support in OZ is a must.  

For now I'm closing this issue as you solved my problem; thanks!
